test/standalone/standalone_test.go

```diff
@@ -0,0 +1,168 @@
+
+import (
+	"embed"
+	"fmt"
+	"io/fs"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	testutils "github.com/envoyproxy/gateway/test/utils"
+	"github.com/envoyproxy/gateway/test/utils/testotel"
+)
+
+//go:embed testdata
+var testdataFS embed.FS
+
+// gatewayBin is the path to the compiled envoy-gateway binary.
+var gatewayBin string
+
+// TestMain builds the binary once for all tests.
+func TestMain(m *testing.M) {
+	var err error
+	if gatewayBin, err = testutils.BuildGoBinaryOnDemand("ENVOY_GATEWAY_BIN", "envoy-gateway", "./cmd/envoy-gateway"); err != nil {
+		fmt.Fprintf(os.Stderr, "failed to build envoy-gateway: %v\n", err)
+		os.Exit(1)
+	}
+	os.Exit(m.Run())
+}
+
+// TestConfig contains template parameters for generating test resources.
+type TestConfig struct {
+	ListenerPort  int
+	CollectorPort int
+	AdminPort     int
+	BackendPort   int
+}
+
+func TestOTELGRPCHeaders(t *testing.T) {
+	ports := testutils.RequireRandomPorts(t, 2)
+	listenerPort := ports[0]
+	adminPort := ports[1]
+
+	// Start a local backend server
+	backendListener, err := net.Listen("tcp", "127.0.0.1:0")
+	require.NoError(t, err)
+	backendPort := backendListener.Addr().(*net.TCPAddr).Port
+	backendServer := &http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(http.StatusOK)
+			_, _ = w.Write([]byte(`{"status": "ok"}`))
+		}),
+	}
+	go func() { _ = backendServer.Serve(backendListener) }()
+	defer backendServer.Close()
+
+	collector, err := testotel.StartGRPCCollector()
+	require.NoError(t, err)
+	defer collector.Close()
+
+	testDir := t.TempDir()
+	cfg := TestConfig{
+		ListenerPort:  listenerPort,
+		CollectorPort: collector.Port(),
+		AdminPort:     adminPort,
+		BackendPort:   backendPort,
+	}
+	generateFromTemplates(t, testDir, "testdata/otel-grpc-headers", cfg)
+
+	cmd := exec.Command(gatewayBin, "server", "--config-path", filepath.Join(testDir, "envoy-gateway.yaml"))
+	cmd.Dir = testDir
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	require.NoError(t, cmd.Start())
+	defer func() {
+		_ = cmd.Process.Kill()
+		_ = cmd.Wait()
+	}()
+
+	// Wait for Envoy to be ready
+	require.Eventually(t, func() bool {
+		resp, err := http.Get(fmt.Sprintf("http://127.0.0.1:%d/", listenerPort))
+		if err != nil {
+			return false
+		}
+		resp.Body.Close()
+		return true
+	}, 30*time.Second, 500*time.Millisecond, "envoy not ready on port %d", listenerPort)
+
+	resp, err := http.Get(fmt.Sprintf("http://127.0.0.1:%d/get", listenerPort))
+	require.NoError(t, err)
+	resp.Body.Close()
+
+	// Verify collector received an access log with the expected format and Authorization header
+	log := collector.TakeLog()
+	require.NotNil(t, log)
+	require.Contains(t, log.Body.GetStringValue(), `HTTP/1.1" 200`)
+	require.Equal(t, "Bearer test-api-key", testotel.GetAttributeString(log.Attributes, "grpc.metadata.authorization"))
```

Comment:
@mathetake @anuraaga you might like this sneaky trick to verify the headers/metadata received by otel. propagate them into fake span attrs